File: src/main/java/io/github/gravetii/controller/GameGridController.java
package io.github.gravetii.controller;

import io.github.gravetii.game.Game;
import io.github.gravetii.game.UserResult;
import io.github.gravetii.model.GameStats;
import io.github.gravetii.model.GridPoint;
import io.github.gravetii.model.GridUnit;
import io.github.gravetii.model.WordResult;
import io.github.gravetii.util.Utils;
import io.github.gravetii.validation.ValidationResult;
import javafx.fxml.FXML;
import javafx.scene.image.ImageView;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.GridPane;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class GameGridController implements FxController {

  private final Game game;
  private final Map<String, ImageView> imgViewMap;
  private final GamePlayValidator validator;
  private final UserResult userResult;

  private GamePlayStyler styler;

  @FXML private GridPane gamePane;

  @FXML private ImageView imgView$0_0;
  @FXML private ImageView imgView$0_1;
  @FXML private ImageView imgView$0_2;
  @FXML private ImageView imgView$0_3;
  @FXML private ImageView imgView$1_0;
  @FXML private ImageView imgView$1_1;
  @FXML private ImageView imgView$1_2;
  @FXML private ImageView imgView$1_3;
  @FXML private ImageView imgView$2_0;
  @FXML private ImageView imgView$2_1;
  @FXML private ImageView imgView$2_2;
  @FXML private ImageView imgView$2_3;
  @FXML private ImageView imgView$3_0;
  @FXML private ImageView imgView$3_1;
  @FXML private ImageView imgView$3_2;
  @FXML private ImageView imgView$3_3;

  public GameGridController(Game game) {
    this.game = game;
    this.imgViewMap = new HashMap<>();
    this.validator = new GamePlayValidator(game.result());
    this.userResult = new UserResult();
  }

  @FXML
  public void initialize() {
    this.imgViewMap.put("imgView$0_0", this.imgView$0_0);
    this.imgViewMap.put("imgView$0_1", this.imgView$0_1);
    this.imgViewMap.put("imgView$0_2", this.imgView$0_2);
    this.imgViewMap.put("imgView$0_3", this.imgView$0_3);
    this.imgViewMap.put("imgView$1_0", this.imgView$1_0);
    this.imgViewMap.put("imgView$1_1", this.imgView$1_1);
    this.imgViewMap.put("imgView$1_2", this.imgView$1_2);
    this.imgViewMap.put("imgView$1_3", this.imgView$1_3);
    this.imgViewMap.put("imgView$2_0", this.imgView$2_0);
    this.imgViewMap.put("imgView$2_1", this.imgView$2_1);
    this.imgViewMap.put("imgView$2_2", this.imgView$2_2);
    this.imgViewMap.put("imgView$2_3", this.imgView$2_3);
    this.imgViewMap.put("imgView$3_0", this.imgView$3_0);
    this.imgViewMap.put("imgView$3_1", this.imgView$3_1);
    this.imgViewMap.put("imgView$3_2", this.imgView$3_2);
    this.imgViewMap.put("imgView$3_3", this.imgView$3_3);
    this.styler = new GamePlayStyler(this.gamePane, this.imgViewMap.values());
  }

  @FXML
  public void onImgViewClick(MouseEvent event) {
    ImageView imgView = (ImageView) event.getSource();
    GridPoint point = Utils.getGridPointFromImageViewLabel(imgView.getId());
    GridUnit unit = game.getGridUnit(point);
    ValidationResult validation = this.validator.validate(unit);
    this.applyStyleAfterValidation(imgView, validation);
  }

  public void rotate() {
    this.styler.rotate();
  }

  private void applyStyleAfterValidation(ImageView imgView, ValidationResult result) {
    if (result == ValidationResult.ALL_INVALID) this.styler.invalidate();
    else if (result == ValidationResult.LAST_INVALID) this.styler.forLastInvalidClick(imgView);
    else this.styler.forValidClick(imgView);
  }

  public Optional<WordResult> validateWordOnBtnClick() {
    Optional<String> word = this.validator.validate();
    if (word.isEmpty()) this.styler.forIncorrectWord();
    Optional<WordResult> result =
        word.flatMap(
            x -> {
              if (this.userResult.exists(x)) {
                this.styler.forRepeatedWord();
                return Optional.empty();
              } else {
                int points = this.game.result().getPoints(x);
                List<GridPoint> seq = this.validator.getSeq();
                WordResult wordResult = new WordResult(x, points, seq);
                this.userResult.add(x, wordResult);
                this.styler.forCorrectWord();
                return Optional.of(wordResult);
              }
            });

    this.validator.reset();
    return result;
  }

  public void revisitUserWord(String word) {
    WordResult result = this.userResult.get(word);
    this.revisit(result);
  }

  public void revisitGameWord(String word) {
    WordResult result = this.game.result().get(word);
    this.revisit(result);
  }

  public Map<String, WordResult> getAllGameWords() {
    return this.game.result().all();
  }

  public Map<String, WordResult> getAllUserWords() {
    return this.userResult.all();
  }

  private void revisit(WordResult result) {
    List<ImageView> imgViews =
        result.getSeq().stream()
            .map(
                point -> {
                  String id = Utils.getImgViewLabelFromGridPoint(point);
                  return this.imgViewMap.get(id);
                })
            .collect(Collectors.toList());

    this.validator.reset();
    this.styler.revisit(imgViews);
  }

  public void endGame() {
    this.styler.applyEndTransition();
    this.imgViewMap.values().forEach(x -> x.setDisable(true));
    this.validator.reset();
    this.styler.invalidate();
    this.styler.rotateOnEnd();
  }

  public GameStats computeStats() {
    return new GameStats(this.game.result(), this.userResult);
  }
}

File: src/main/kotlin/io/github/gravetii/controller/GameGridController.kt
package io.github.gravetii.controller

import io.github.gravetii.game.Game
import io.github.gravetii.game.UserResult
import io.github.gravetii.model.GameStats
import io.github.gravetii.model.GridPoint
import io.github.gravetii.model.WordResult
import io.github.gravetii.validation.ValidationResult
import javafx.fxml.FXML
import javafx.scene.image.ImageView
import javafx.scene.input.MouseEvent
import javafx.scene.layout.GridPane

class GameGridController(private val game: Game) : FxController {

    private val imgViewMap: MutableMap<String, ImageView> = mutableMapOf()
    private val validator = GamePlayValidator(game.result)
    private val userResult = UserResult()

    private lateinit var styler: GamePlayStyler

    @FXML
    private lateinit var gamePane: GridPane

    @FXML
    private lateinit var imgView_0_0: ImageView

    @FXML
    private lateinit var imgView_0_1: ImageView

    @FXML
    private lateinit var imgView_0_2: ImageView

    @FXML
    private lateinit var imgView_0_3: ImageView

    @FXML
    private lateinit var imgView_1_0: ImageView

    @FXML
    private lateinit var imgView_1_1: ImageView

    @FXML
    private lateinit var imgView_1_2: ImageView

    @FXML
    private lateinit var imgView_1_3: ImageView

    @FXML
    private lateinit var imgView_2_0: ImageView

    @FXML
    private lateinit var imgView_2_1: ImageView

    @FXML
    private lateinit var imgView_2_2: ImageView

    @FXML
    private lateinit var imgView_2_3: ImageView

    @FXML
    private lateinit var imgView_3_0: ImageView

    @FXML
    private lateinit var imgView_3_1: ImageView

    @FXML
    private lateinit var imgView_3_2: ImageView

    @FXML
    private lateinit var imgView_3_3: ImageView

    @FXML
    fun initialize() {
        imgViewMap["imgView_0_0"] = imgView_0_0
        imgViewMap["imgView_0_1"] = imgView_0_1
        imgViewMap["imgView_0_2"] = imgView_0_2
        imgViewMap["imgView_0_3"] = imgView_0_3
        imgViewMap["imgView_1_0"] = imgView_1_0
        imgViewMap["imgView_1_1"] = imgView_1_1
        imgViewMap["imgView_1_2"] = imgView_1_2
        imgViewMap["imgView_1_3"] = imgView_1_3
        imgViewMap["imgView_2_0"] = imgView_2_0
        imgViewMap["imgView_2_1"] = imgView_2_1
        imgViewMap["imgView_2_2"] = imgView_2_2
        imgViewMap["imgView_2_3"] = imgView_2_3
        imgViewMap["imgView_3_0"] = imgView_3_0
        imgViewMap["imgView_3_1"] = imgView_3_1
        imgViewMap["imgView_3_2"] = imgView_3_2
        imgViewMap["imgView_3_3"] = imgView_3_3
        styler = GamePlayStyler(gamePane, imgViewMap.values)
    }

    private fun String.getImageViewLabel(): GridPoint {
        val tokens = split("_")
        return GridPoint(tokens[1].toInt(), tokens[2].toInt())
    }

    private fun applyStyleAfterValidation(imgView: ImageView, result: ValidationResult) {
        when (result) {
            ValidationResult.ALL_INVALID -> styler.invalidate()
            ValidationResult.LAST_INVALID -> styler.forLastInvalidClick(imgView)
            else -> styler.forValidClick(imgView)
        }
    }

    private fun revisit(result: WordResult) {
        val imgViews = result.seq.map {
            val imgView = "imgView_${it.x}_${it.y}"
            imgViewMap[imgView]!!
        }

        validator.reset()
        styler.revisit(imgViews)
    }

    @FXML
    fun onImgViewClick(event: MouseEvent) {
        val imgView = event.source as ImageView
        val point = imgView.id.getImageViewLabel()
        val unit = game.getGridUnit(point)
        applyStyleAfterValidation(imgView, validator.validate(unit))
    }

    fun rotate() = styler.rotate()

    fun validateWordOnBtnClick(): WordResult? {
        val word = validator.validate()
        val result = if (word == null) {
            styler.forIncorrectWord()
            null
        } else if (userResult.exists(word)) {
            styler.forRepeatedWord()
            null
        } else {
            val points = game.result.getPoints(word)
            val seq = validator.getSeq()
            val wordResult = WordResult(word, points, seq)
            userResult.add(word, wordResult)
            styler.forCorrectWord()
            wordResult
        }

        validator.reset()
        return result
    }

    fun revisitUserWord(word: String) = revisit(userResult.get(word))

    fun revisitGameWord(word: String) = revisit(game.result.get(word))

    fun getAllGameWords(): Map<String, WordResult> = game.result.all()

    fun getAllUserWords(): Map<String, WordResult> = userResult.all()

    fun endGame() {
        styler.applyEndTransition()
        imgViewMap.values.forEach { it.isDisable = true }
        validator.reset()
        styler.rotateOnEnd()
    }

    fun computeStats(): GameStats = GameStats(game.result, userResult)

}